src/tdyio.c

```diff
@@ -11,43 +12,107 @@ PetscErrorCode TDyIOCreate(TDyIO *_io) {
 
   io->io_process = PETSC_FALSE;
   io->print_intermediate = PETSC_FALSE;  
-  strcpy(io->exodus_filename, "out.exo");
   io->num_vars = 1;
   strcpy(io->zonalVarNames[0], "Soln");
-  io->format = PetscViewerASCIIFormat;
+  io->format = NullFormat;
   io->num_times = 0;
     
   PetscFunctionReturn(0);
 }
 
-PetscErrorCode TDyIOSetMode(TDyIO io, TDyIOFormat format){
+PetscErrorCode TDyIOSetMode(TDy tdy, TDyIOFormat format){
   PetscFunctionBegin;
-  io->format=format;
+  PetscErrorCode ierr;
+  
+  tdy->io->format = format;
+  int num_vars = tdy->io->num_vars;
+  DM dm = tdy->dm;
+  char *zonalVarNames[1];
+
+  PetscInt dim,istart,iend,numCell,numVert,numCorner;
+
+  zonalVarNames[0] = tdy->io->zonalVarNames[0];
+
+  if (tdy->io->format == ExodusFormat) {
+    strcpy(tdy->io->filename, "out.exo");
+    char *ofilename = tdy->io->filename;
+    ierr = TdyIOInitializeExodus(ofilename,zonalVarNames,dm,num_vars);CHKERRQ(ierr);
+  }
+  else if (tdy->io->format == HDF5Format) {
+    strcpy(tdy->io->filename, "out.h5");
+    char *ofilename = tdy->io->filename;
+    numCorner = TDyGetNumberOfCellVertices(dm);
+    ierr = DMGetDimension(dm,&dim);CHKERRQ(ierr); 
+    ierr = DMPlexGetHeightStratum(dm,3,&istart,&iend);CHKERRQ(ierr); //just for box example?
+    numVert = iend-istart;
+    ierr = DMPlexGetHeightStratum(dm,0,&istart,&iend);CHKERRQ(ierr);
+    numCell = iend-istart;
+    
+      
+    ierr = TdyIOInitializeHDF5(ofilename,dm);CHKERRQ(ierr);
+    ierr = TDyIOWriteXMFHeader(numCell,dim,numVert,numCorner);CHKERRQ(ierr);
+  }
+    
   PetscFunctionReturn(0);
 }
 
 PetscErrorCode TDyIOWriteVec(TDy tdy){
   PetscErrorCode ierr;
-  char *zonalVarNames[1];
- 
-  char *ofilename = tdy->io->exodus_filename;
+  PetscInt numCell,istart,iend;
+  
   int num_vars = tdy->io->num_vars;
   Vec v = tdy->solution;
   DM dm = tdy->dm;
   PetscReal time = tdy->ti->time;
-  zonalVarNames[0] = tdy->io->zonalVarNames[0];
-  
+ 
   if (tdy->io->format == PetscViewerASCIIFormat) {
-    TDyIOPrintVec(v, time);
+    ierr = TDyIOWriteAsciiViewer(v, time);CHKERRQ(ierr);
   }
-  if (tdy->io->format == ExodusFormat){
-    if (tdy->io->num_times == 0) {
-      TdyIOInitializeExodus(ofilename,zonalVarNames,dm,num_vars);
-      ierr = PetscObjectSetName((PetscObject) v,  "Soln");CHKERRQ(ierr);
-    }
-    TdyIOAddExodusTime(ofilename,time,tdy->io);
-    TdyIOWriteExodusVar(ofilename,v,tdy->io);
+  else if (tdy->io->format == ExodusFormat) {
+    char *ofilename = tdy->io->filename;
+
+    ierr = TdyIOAddExodusTime(ofilename,time,tdy->io);CHKERRQ(ierr);
+    ierr = TdyIOWriteExodusVar(ofilename,v,tdy->io);CHKERRQ(ierr);
+  }
+  else if (tdy->io->format == HDF5Format) {
```

Comment:
@leorosie, I believe here is the location where you want to rearrange the vector following the code at [tdycoremesh.c#L422](https://github.com/TDycores-Project/TDycore/blob/master/src/mesh/tdycoremesh.c#L422) 

```
ierr = DMGetUseNatural(dm, &useNatural); CHKERRQ(ierr);
if (useNatural) {
   // Will reorder the vector
   // 1. Create a natural vector
   // 2. Use DMPlexGlobalToNaturalBegin/End to reorder the vector
   // 3. Send the natural vector to TdyIOWriteHDF5Var
}
```